src/ray/core_worker/reference_count.h

```diff
@@ -113,20 +126,14 @@ class ReferenceCounter {
     /// if we do not know the object's owner (because distributed ref counting
     /// is not yet implemented).
     absl::optional<std::pair<TaskID, rpc::Address>> owner;
+    /// Callback that will be called when this ObjectID no longer has references.
+    std::function<void(const ObjectID &)> on_delete;
   };
 
-  /// Helper function with the same semantics as AddReference to allow adding a reference
-  /// while already holding mutex_.
-  void AddLocalReferenceInternal(const ObjectID &object_id)
-      EXCLUSIVE_LOCKS_REQUIRED(mutex_);
-
-  /// Recursive helper function for decreasing reference counts. Will recursively call
-  /// itself on any dependencies whose reference count reaches zero as a result of
-  /// removing the reference.
-  ///
-  /// \param[in] object_id The object to to decrement the count for.
-  /// \param[in] deleted List to store objects that hit zero ref count.
-  void RemoveReferenceRecursive(const ObjectID &object_id, std::vector<ObjectID> *deleted)
+  /// Helper method to delete an entry from the reference map and run any necessary
+  /// callbacks. Assumes that the entry is in object_id_refs_.
```

Comment:
Maybe it's obvious but could also add a note that this invalidates the iterator.